pyteal/ast/abi/util.py

```diff
@@ -499,3 +499,47 @@ def type_specs_from_signature(sig: str) -> tuple[list[TypeSpec], Optional[TypeSp
         return_type = type_spec_from_algosdk(sdk_method.returns.type)
 
     return [type_spec_from_algosdk(arg.type) for arg in sdk_method.args], return_type
+
+
+def type_spec_is_assignable(a: TypeSpec, b: TypeSpec) -> bool:
+    from pyteal.ast.abi import (
+        TupleTypeSpec,
+        ArrayTypeSpec,
+        StaticArrayTypeSpec,
+        DynamicArrayTypeSpec,
+    )
+
+    match a, b:
+        case TupleTypeSpec(), TupleTypeSpec():
```

Comment:
For both case 1 and 2, I think we should all them, and I extended the method such that both cases can pass, in both ways of assignments (a, b or b, a).

For case 3, I add a new testcase for assigning between different `namedTuples`. Let me know if I should extend them.

Reply:
After more thought, I propose this rule: give types `a` and `b`, a value of type `a` is safe to view as type `b` (i.e. `type_spec_is_assignable(a,b)`) if and only if the following properties are true:
* Types `a` and `b` are encoded identically
* And type `b` is as general or more general than type `a`

So that would imply the following:

| Type `a` | Type `b` | `type_spec_is_assignable(a,b)` | reason |
| -------- | -------- | ------------------------------ | ------ |
| `DynamicArray[Byte]` | `DynamicBytes` | `true` | `DynamicBytes` is as general as `DynamicArray[Byte]` |
| `DynamicBytes` | `DynamicArray[Byte]` | `true` | `DynamicArray[Byte]` is as general as `DynamicBytes` |
| `StaticArray[Byte,N]` | `StaticBytes[N]` | `true` | `StaticBytes[N]` is as general as `StaticArray[Byte,N]` |
| `StaticBytes[N]` | `StaticArray[Byte,N]` | `true` | `StaticArray[Byte,N]` is as general as `StaticBytes[N]` |
| `String` | `DynamicBytes` | `true` | `DynamicBytes` is more general than `String` |
| `DynamicBytes` | `String` | `false` | `String` is more specific than `DynamicBytes` |
| `Address` | `StaticBytes[32]` | `true` | `StaticBytes[32]` is more general than `Address` |
| `StaticBytes[32]` | `Address` | `false` | `Address` is more specific than `StaticBytes[32]` |

Does this make sense?

Reply:
Good good, that makes sense to me. I just made the change to implementation and testcase.

Next up is to have a full coverage over abi object reflected by the discussion on last Friday (2022/09/30).

Reply:
Yes, I will take a pass and try to get it merged into my implementation.